wip: prost & tonic

freeformstu opened the pull request.

freeformstu committed:
Initial prost & tonic setup.

abrisco commented on third_party/prost/BUILD.bazel:
This all should definitely be under `proto` and be hidden behind the same macros.

abrisco commented on third_party/prost/private/protoc_wrapper.rs:
```diff
@@ -0,0 +1,404 @@
+        .collect())
+}
+
+/// Generate a lib.rs file with all prost/tonic outputs embeeded in modules which
```
Apparently there's some rustdoc guide where all docs start with `Returns ...` and the arguments section looks like
```
# Arguments

* `arg1` - docs
* `arg2` - docs
```
Should probably use this here and document `package_infos`.

abrisco commented on third_party/prost/private/protoc_wrapper.rs:
````diff
@@ -0,0 +1,404 @@
+///     }
+/// }
+/// ```
+fn generate_lib_rs(package_infos: &BTreeMap<String, PathBuf>) -> Result<String> {
````
This function felt like an interview question that I had to answer after many drinks, does it seem ok?

abrisco commented on third_party/prost/private/protoc_wrapper.rs:
```diff
@@ -0,0 +1,404 @@
+    .context("Failed to write package info")?;
+
+    // Finally run rustfmt on the output lib.rs file
+    let runfile = env!("RUSTFMT")
```
I think `rustfmt` should actually come from the action. So I'd make this optional and instead try to parse yet another argument in the list above. Maybe `--rustfmt=bazel-bin/path/to/rustfmt`? This way in the future it can be an optional toolchain in case folks are trying to run bazel on weird hosts.

abrisco commented on third_party/prost/private/tests/hello_tonic/helloworld_messages.proto:
```diff
@@ -0,0 +1,12 @@
+syntax = "proto3";
+package helloworld_messages;
```
We should actually add many duplicates of this example but with a variety of different kinds of package names
`helloworld.messages`
`hello.world.messages`
Make a sub directory with many layers of things
```

.
\_ foo/
      \_ proto files here...
\_ bar/
      \_ baz/
      |     \_ proto files here...
      \_ proto files here...
```
This way variations of `proto_library` are covered since that's the source of truth for the generated crate.

abrisco commented on third_party/prost/private/tests/hello_tonic/helloworld_server.rs:
I never got around to it but I think it'd be good to actually have a test that actually utilized the generated contents and not just compiled with them.

abrisco commented on third_party/prost/private/prost.bzl:
```diff
@@ -0,0 +1,334 @@
+load("@rules_rust//rust:defs.bzl", "rust_common")
+
+# buildifier: disable=bzl-visibility
+load("@rules_rust//rust/private:providers.bzl", "CrateInfo", "DepInfo", "DepVariantInfo")
```
These can all be accessed through `rust_common` now.

abrisco commented on third_party/prost/private/prost.bzl:
```diff
@@ -0,0 +1,334 @@
+                allow_single_file = True,
+                cfg = "exec",
+            ),
+            "_prost_proto_toolchain": attr.label(
```
I would actually promote this to a proper toolchain. Make a `@rulest_rust//rust/proto/prost:toolchain_type` maybe and that's what users will register. I would then construct a `ProtoLangToolchainInfo` from that in the toolchain implementation to be able to use the same API. This would probably allow you to consolidate this aspect down to one where the toolchain determines if it's a tonic or pure prost library you want to make.

abrisco commented on third_party/prost/private/BUILD.bazel:
```diff
@@ -0,0 +1,79 @@
+        ] if kind == "tonic" else []),
+        rustc_env = dict({
+            "PROST_EXTENSION": PROST_EXTENSION,
+            "PROTOC": "$(rootpath @com_google_protobuf//:protoc)",
```
The `protoc` and `protoc-gen-tonic` plugin should probably come from the rule. This process wrapper could stand to be dumber

abrisco commented on third_party/prost/private/protoc_wrapper.rs:
```diff
@@ -0,0 +1,404 @@
+    Ok(format!(".{}={}", package, module))
+}
+
+#[derive(Debug, Parser)]
```
I actually think this should not use `clap` so it's a lighter weight tool for consumers of the rule.

cameron-martin committed:
Exclude .tmp_git_root from globs (#1948)

* Exclude .tmp_git_root from globs

This directory contains the original git repo when the crate is from git. Including this directory currently makes these rules re-run whenever the repository rule re-runs, although this is fixed by https://github.com/bazelbuild/bazel/pull/18271. Even after this fix, excluding this directory avoids depending on unnecessary files.

Fixes #1927.

* Regenerate vendored crates

UebelAndre committed:
Fix flaky coverage test in CI (#2028)

illicitonion committed:
Re-enable zig example on CI (#2030)

rickvanprim committed:
fix: `crate_type` more accurately corresponds to CC linking actions (#1975)
